Upstream contributing

pbartholomew08 opened the pull request.

pbartholomew08 committed:
Update the CONTRIBUTING guide

To prevent circular commits between github and gitlab, external
contributions should target the upstreaming branch, rather than
develop directly. Upstreaming will then merge into develop in the
upstream repository

pbartholomew08 committed:
Fix typo

pbartholomew08 committed:
Update CONTRIBUTING.md

Clarify the steps contributors should take, and that core developer should take.

pbartholomew08 committed:
Highlight steps core developers should take to assess proposed contributions

pbartholomew08 committed:
Clarify steps for starting and accepting contributions

In particular how the upstreaming branch should be kept in sync with
develop

pbartholomew08 committed:
Add note for external contributions that affect tests

We should take additional care in reviewing such changes, and have
at least two approvals.

justsz commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,96 @@
+
+### Before you start
+
+Although the ASiMoV-CCS developers will try to keep `upstreamin` synchronised with `develop` it is
```
typo - "upstreamin"

pbartholomew08 replied:
Done

justsz commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,96 @@
 
-Before a contribution is merged it will be reviewed by one of the ASiMoV-CCS developers, in particular to check the coding standard is followed and to verify the full test suite must still pass.
+Before a contribution is merged it will be reviewed by one of the ASiMoV-CCS developers, in
+particular to check the coding standard is followed and to verify the full test suite must still
```
do you mean "full test suite still passes"?

pbartholomew08 replied:
Done

justsz commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,96 @@
+
+### Keeping up to date
+
+We recommend periodically pulling the `origin/develop` branch and merging into your working branch
+if the development of your contribution continues for some time, this will ensure that it does not
+stray too far from the eventual merge target.
```
Is this a merge instead of a rebase on purpose?

justsz replied:
Nevermind - we should remove most of this section and just tell the user to write their contribution against `develop` and only mention `upstreaming` when it comes to the Pull Request.

pbartholomew08 replied:
Done?

justsz replied:
Yep.

pbartholomew08 committed:
Correct typo

pbartholomew08 committed:
Correct poor phrasing

pbartholomew08 committed:
Remove most references to upstreaming branch

This is mostly an internal/external-facing issue and contributions
should work against the develop branch, only the pull request should
target the upstreaming branch (and this can be resolved by CCS developers
if a mistake is made)

justsz commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,94 @@
+
+### Keeping up to date
+
+We recommend periodically pulling the `origin/develop` branch and merging into your working branch
+if the development of your contribution continues for some time, this will ensure that it does not
+stray too far from the eventual merge target.
```
I think this should also be a rebase instead of a merge.

pbartholomew08 replied:
Done

pbartholomew08 committed:
Addressing comment to recommend rebase over merge

justsz reviewed (approved):
LGTM!

Nanoseb commented on CONTRIBUTING.md:
````diff
@@ -1,6 +1,95 @@
+To do so we recommend using `git rebase` to keep the git history clean, assuming that the main
+ASiMoV-CCS repository is `origin` and your branch is called `my-branch`:
+```
+git checkout develop
````
maybe pedantic, but `git switch` is there for that now. Same a couple of line below.

Also why not just 
```
git switch my-branch
git pull origin/develop
```

pbartholomew08 replied:
but will that cause their branch to check `origin/develop`? I don't want to get in a mess with users trying to push back to the `origin/develop` branch. Also this is assuming they've already created a branch and are keeping it up to date

Nanoseb replied:
It pulls from `origin/develop`, but doesn't change the branch name or remote, any future push will be to `my-branch` etc. It is just a way to update `my-branch` with the latest develop in a single line. I believe it is equivalent to what you had before, but I may be wrong.
Anyway, sorry for the confusion, real no need to change that part if the current version is working.

Nanoseb replied:
(and yes it assumes `my-branch` exists, but so does `git checkout my-branch` right? It would need `-b` or `-B` to create it)

pbartholomew08 replied:
yes, my text is meant for _after_ they've created their branch and keeping it up to date. I think I will leave as is as I don't want to write a git tutorial

Nanoseb commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,95 @@
+  `develop` by rebasing (see above instructions) and push the updated `upstreaming` branch to GitHub
+1. When a pull request is opened check it is targetting the `upstreaming` branch, change the target
+   branch if necessary.
+2. Ensure that the `upstreaming` branch is up to date by rebasing and pushing to GitHub if
```
This is a duplicate of line 81

pbartholomew08 replied:
Removed line 81 to clarify (original instructions had them work against upstreaming, now it is only important when they come to merge into upstreaming)

Nanoseb reviewed (commented):
two minor things that may not even need to be addressed

pbartholomew08 committed:
Remove double instruction for CCS devs to update upstreaming branch

This was intended to ensure that external contributions pull from an up to date branch, but new
guidance is to pull from develop and contribute to upstreaming. It is at this point it is important
to ensure upstreaming is up to date
